mp4parse/src/lib.rs

```diff
@@ -689,6 +721,180 @@ impl MediaContext {
     }
 }
 
+#[derive(Debug, Default)]
+pub struct AvifContext {
+    /// The collected data indicated by the `pitm` box, See ISO 14496-12:2015 § 8.11.4
+    pub primary_item: Vec<u8>,
+}
+
+impl AvifContext {
+    pub fn new() -> Self {
+        Default::default()
+    }
+}
+
+/// A Media Data Box
+/// See ISO 14496-12:2015 § 8.1.1
+struct Mdat {
+    /// Offset of `data` from the beginning of the file. See ConstructionMethod::FileOffset
+    offset: u64,
+    data: Vec<u8>,
+}
+
+impl Mdat {
+    /// Check whether the beginning of `extent` is within the bounds of the `Mdat`.
+    /// We assume extents to not cross box boundaries. If so, this will cause an error
+    /// in `read_extent`.
+    fn contains_extent(&self, extent: &ExtentRange) -> bool {
+        if self.offset <= extent.start() {
+            let start_offset = extent.start() - self.offset;
+            start_offset < self.data.len().to_u64()
+        } else {
+            false
+        }
+    }
+
+    /// Check whether `extent` covers the `Mdat` exactly.
+    fn matches_extent(&self, extent: &ExtentRange) -> bool {
+        if self.offset == extent.start() {
+            match extent {
+                ExtentRange::WithLength(range) => {
+                    if let Some(end) = self.offset.checked_add(self.data.len().to_u64()) {
+                        end == range.end
+                    } else {
+                        false
+                    }
+                }
+                ExtentRange::ToEnd(_) => true,
+            }
+        } else {
+            false
+        }
+    }
+
+    /// Copy the range specified by `extent` to the end of `buf` or return an error if the range
+    /// is not fully contained within `Mdat`.
+    fn read_extent(&mut self, extent: &ExtentRange, buf: &mut Vec<u8>) -> Result<()> {
+        let start_offset = extent.start().checked_sub(self.offset).expect("mdat does not contain extent");
+        let slice = match extent {
+            ExtentRange::WithLength(range) => {
+                let range_len = range.end.checked_sub(range.start).expect("range start > end");
+                let end = start_offset.checked_add(range_len).expect("extent end overflow");
+                self.data.get(start_offset.try_into()?..end.try_into()?)
+            }
+            ExtentRange::ToEnd(_) => {
+                self.data.get(start_offset.try_into()?..)
+            }
+        };
+        let slice = slice.ok_or(Error::InvalidData("extent crosses box boundary"))?;
+        buf.extend_from_slice(slice);
```

Comment:
Will need to consider fallible allocation here.

Reply:
It looks pretty straightforward to add a `FallibleVec::try_extend_from_slice` to [`mp4parse_fallible`](https://github.com/mozilla/mp4parse_fallible). Would you like me to add that before merging this?

I notice there are a couple other [places where `extend_from_slice` is used](https://github.com/mozilla/mp4parse-rust/search?q=extend_from_slice&unscoped_q=extend_from_slice). Do those also need updating, or is this case fundamentally different (aside from the slice being likely larger)?

Reply:
The small ones are fine, but it'd be worth investigating the potential sizes (or source of sizes, e.g. from untrusted data) of some of them.

Generally, if the allocation could be "large" or the size of the allocation is user controlled without limits, use a fallible allocator.  If the allocation is "small", trusted code controls the size, it's a user-supplied size but is known to be small, or it's a situation where continuing after an allocation failure is impossible, it's fine to use the infallible allocators.  The concept of "large" and "small" are somewhat nebulous and there are edge cases to consider (e.g. if you've got a large existing allocation and try to grow it by a small amount, it can fail), so it's not entirely straightforward.

I tried to find good documentation on Gecko's policy (which we try to follow here), but only found https://developer.mozilla.org/en-US/docs/Mozilla/Infallible_memory_allocation#Choosing_a_memory_allocator which is fairly terse.

Reply:
It looks like there are 4 existing instances:
1. [`pssh_box.extend_from_slice(b"pssh");`](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse/src/lib.rs#L1049): definitely small
1. [`pssh_data.extend_from_slice(pssh.system_id.as_slice());`](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse_capi/src/lib.rs#L1383): in practice, [exactly 16 bytes](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse/src/lib.rs#L1030)
1. [`pssh_data.extend_from_slice(data_len.as_slice());`](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse_capi/src/lib.rs#L1384): [4 bytes](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse_capi/src/lib.rs#L1380)
1. [`pssh_data.extend_from_slice(pssh.box_content.as_slice());`](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse_capi/src/lib.rs#L1385): doesn't seem to be inherently bounded; [size based on untrusted input](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse/src/lib.rs#L1022).

I'll just change the last one and add a comment.